# Datenbanken

## Praxisbeispiel

Informationen sind in der Praxis meist über verschiedene Systeme und Tabellen verteilt und müssen oft zusammengeführt verarbeitet werden, um daraus anwendbares Wissen abzuleiten.

Die Hansestadt Rostock bietet zum Beispiel viele relevante Daten für Bau- und Umweltingenieure auf dem [OpenData Portal Rostock](https://www.opendata-hro.de).

Dort erhalten wir zum Beispiel:
- die [Liste aller Baustellen](https://www.opendata-hro.de/dataset/baustellen)
- die [Liste aller Strassen und Adressen](https://www.opendata-hro.de/dataset/adressenliste)
- die [Liste aller Gemeinden](https://www.opendata-hro.de/dataset/gemeinden_mecklenburg-vorpommern) 
<!-- - die [Liste aller Bodenrichtwerte](https://www.opendata-hro.de/dataset/bodenrichtwerte_2022)

Die Bevölkerungsdichte der Bezirke wiederum erhalten wir auf der Landesseite (https://www.laiv-mv.de/Statistik/Zahlen-und-Fakten/Gesellschaft-&-Staat/Bevoelkerung/). 

# Siehe https://www.opendata-hro.de/dataset/bodenrichtwerte_2022
with urllib.request.urlopen(f'https://geo.sv.rostock.de/download/opendata/bodenrichtwerte_2022/bodenrichtwerte_2022.csv') as f:
    bodenrichtwerte=pd.read_csv(f, usecols=["gesl", "genu", "ortst", "wnum", "gema", "flae", "typ", "brw", "nuta", "entw", "geometrie_wkt"])
bodenrichtwerte.head()

# Siehe https://www.laiv-mv.de/Statistik/Zahlen-und-Fakten/Gesellschaft-&-Staat/Bevoelkerung/
with urllib.request.urlopen(f'https://www.laiv-mv.de/static/LAIV/Statistik/Dateien/Publikationen/A%20I%20Bev%C3%B6lkerungsstand/A123/2022/A123%202022%2021.xlsx') as f:
    bevoelkerung=pd.read_excel(f.read(), sheet_name="2", header=1, usecols=range(7), skiprows=range(2,9)) # wähle Blatt 2, die ersten 7 Spalten und überspringe Zeilen 2-9
bevoelkerung.columns=[c.replace("-\n", "").replace("\n", "") for c in bevoelkerung.columns] # entferne Zeilenumbrüche in den Spaltennamen
bevoelkerung.head()

-->

Wie in den letzten Beispielen gezeigt können wir uns die Daten auch sehr einfach in Python herunter laden und anzeigen. Dafür nutzen wir das CSV Format das auf den Webseiten angeboten wird und die Python Pakete `urllib` und `pandas`. Wir öffnen zuerst die Datei im Internet mit `urllib.request.urlopen` und laden sie dann in Pandas als Tabelle mit `pd.read_csv(f)`. Zuletzt zeigen wir die ersten 10 Zeilen der Tabelle mit `.head()` an.

In [1]:
import urllib.request
import pandas as pd

In [2]:
# Siehe https://www.opendata-hro.de/dataset/baustellen
with urllib.request.urlopen(f'https://geo.sv.rostock.de/download/opendata/baustellen/baustellen.csv') as f:
    baustellen = pd.read_csv(f, usecols=["latitude", "longitude", "strasse_schluessel", "sparte", "von", "nach", "baubeginn", "bauende", "verkehrsbeeintraechtigungen", "baumassnahme"], parse_dates=["baubeginn", "bauende"])
baustellen.head()

latitude  longitude  strasse_schluessel            sparte  von nach  \
0  54.092633  12.147336                 530           Hochbau    5   6a   
1  54.067580  12.076324                5880           Hochbau  NaN  NaN   
2  54.143706  12.068219                6050           Hochbau   1c   1e   
3  54.083986  12.132370                 970  Gebäudesanierung   89   88   
4  54.077593  12.116732                2030           Hochbau   24  NaN   

                   baubeginn                    bauende  \
0  2022-09-01 00:00:00+02:00  2024-01-01 00:00:00+01:00   
1  2022-09-02 00:00:00+02:00  2023-06-03 00:00:00+02:00   
2  2022-03-07 09:00:00+01:00  2024-05-31 18:00:00+02:00   
3  2022-12-07 00:00:00+01:00  2023-07-01 00:00:00+02:00   
4  2022-01-22 00:00:00+01:00  2023-02-04 00:00:00+01:00   

                         verkehrsbeeintraechtigungen  \
0  Sicherungsmaßnahmen entlang der Straße, Sicher...   
1  Sicherungsmaßnahmen entlang der Straße, Sicher...   
2  Sicherungsmaßnahmen entlang der Straße, Sicher...   
3                                                NaN   
4                                                NaN   

                                baumassnahme  
0                                    Hochbau  
1  Erschließung B-Plan Wohngebiet Kiefernweg  
2     Neubau Mehrfamilienhaus mit Tiefgarage  
3                                        NaN  
4                                        NaN

In [3]:
# Siehe https://geo.sv.rostock.de/download/opendata/adressenliste
with urllib.request.urlopen(f'https://geo.sv.rostock.de/download/opendata/adressenliste/adressenliste.csv') as f:
    adressenliste=pd.read_csv(f, usecols=["gemeinde_schluessel", "gemeindeteil_name", "strasse_name", "strasse_schluessel", "hausnummer", "hausnummer_zusatz", "postleitzahl"])
adressenliste.head()

gemeinde_schluessel gemeindeteil_name         strasse_name  \
0         130030000000         Gehlsdorf             Blockweg   
1         130030000000           Schmarl        Industriestr.   
2         130030000000           Schmarl        Industriestr.   
3         130030000000        Evershagen  Bertolt-Brecht-Str.   
4         130030000000        Evershagen      An der Jägerbäk   

   strasse_schluessel  hausnummer hausnummer_zusatz  postleitzahl  
0                1430           3               NaN         18147  
1                4010           8               NaN         18069  
2                4010          11               NaN         18069  
3                1340          17                 a         18106  
4                 680           7               NaN         18069

In [4]:
# Siehe https://www.opendata-hro.de/dataset/gemeinden_mecklenburg-vorpommern
with urllib.request.urlopen(f'https://geo.sv.rostock.de/download/opendata/gemeinden_mecklenburg-vorpommern/gemeinden_mecklenburg-vorpommern.csv') as f:
    gemeinden=pd.read_csv(f, usecols=["kreis_name", "kreis_schluessel", "gemeindeverband_name", "gemeindeverband_schluessel", "gemeinde_name", "gemeinde_schluessel", "WKT"])
gemeinden.head()

WKT             kreis_name  \
0  MULTIPOLYGON (((11.9096631786872 53.6385042791...    Ludwigslust-Parchim   
1  MULTIPOLYGON (((11.8594842394746 53.6614357447...    Ludwigslust-Parchim   
2  MULTIPOLYGON (((11.3738758252458 53.1928733880...    Ludwigslust-Parchim   
3  MULTIPOLYGON (((12.2091673297715 53.6031013139...    Ludwigslust-Parchim   
4  MULTIPOLYGON (((13.7101796941594 53.4791202040...  Vorpommern-Greifswald   

   kreis_schluessel            gemeindeverband_name  \
0             13076  Amt Sternberger Seenlandschaft   
1             13076  Amt Sternberger Seenlandschaft   
2             13076                      Amt Grabow   
3             13076          Amt Goldberg-Mildenitz   
4             13075    Strasburg (Uckermark), Stadt   

   gemeindeverband_schluessel                 gemeinde_name  \
0                   130765664                         Dabel   
1                   130765664              Sternberg, Stadt   
2                   130765657                        Eldena   
3                   130765656                   Neu Poserin   
4                   130750130  Strasburg (Uckermark), Stadt   

   gemeinde_schluessel  
0         130765664026  
1         130765664128  
2         130765657037  
3         130765656104  
4         130750130130

Wenn wir die Tabellen vergleichen, so stellen wir fest, dass alle eine etwas andere Struktur haben und meist nur einen Teil der Informationen erhalten. Wollen wir zum Beispiel wissen was die Bevölkerungsdichte in den Stadtbezirken mit Baustellen ist, so lässen sich dieses Wissen nicht sofort ableiten.

## Tabellen in SQLite speichern mit Pandas

Um diese Daten langfristig zu speichern wollen wir uns eine Datenbank anlegen. Da die Daten ja bereits als Tabellen vorliegen, nutzen wir eine relationale Datenbank. Eine sehr einfache, lokale relationale Datenbank die ohne Server auskommt ist SQLite. Sie ist bereits in Python enthalten. Wir erzeugen uns eine neue Datenbank indem wir eine Verbindung zu einer neuen Datenbankdatei anlegen.

In [5]:
import sqlite3

# Create a SQL connection to our SQLite database
con = sqlite3.connect("opendatahro.sqlite")

Wir schauen uns später an wie wir Tabellen direkt mit SQL erzeugen. Pandas hat diese Funktion bereits eingebaut, so dass wir unsere Tabelle direkt in der SQLite Datenbank speichern können.

In [6]:
baustellen.to_sql("baustellen", con, if_exists="replace")
adressenliste.to_sql("adressenliste", con, if_exists="replace")
gemeinden.to_sql("gemeinden", con, if_exists="replace")

726

## Daten aus Tabellen abfragen mit `SELECT`

SQL ist die Standartsprache um mit relationalen Datenbanken zu arbeiten. Sie bietet verschiedene Befehle um Tabellen zu definieren, Daten in ihnen zu speichern als auch die Daten abzufragen. Letzteres wird mit dem `SELECT` Befehl gemacht. SQL orientiert sich dabei etwas an natürlicher Sprache. Man fragt in SQL quasi nach den Daten aus (`FROM`) einer bestimmten Tabelle . Um zum Beispiel alle Spalten aus einer Tabelle abzufragen nutzen wir den Befehl mit dem Platzhalter `*` (=alle Spalten).

In [7]:
sql='SELECT * FROM baustellen;'

Um den SQL Befehl an die Datenbank zu senden, erzeugen wir aus unserer Datenbankverbindung `con` einen neuen Cursor `cur`. Mit `cur.execute(sql)` führen wir den SQL Befehl aus und die Datenbank sendet und die Datenbank fängt an uns alle Ergebnisse der Anfrage Zeile für Zeile zurück zu senden. Der Cursor zeigt dabei immer auf die aktuelle Zeile. Damit will man vermeiden alle Daten, welche sehr (sehr) viele Zeilen sein können, auf einmal zu senden und damit ggf. das Netzwerk oder das Programm zu überlasten.

In [8]:
cur = con.cursor()

rows=cur.execute(sql)
for row in rows:
    print(row)
print(type(row))

(0, 54.0926332355401, 12.1473363033309, 530, 'Hochbau', '5', '6a', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Sicherungsmaßnahmen entlang der Straße, Sicherungsmaßnahmen entlang des Gehwegs, Verkehrsraumeinschränkungen', 'Hochbau')
(1, 54.0675802011562, 12.0763242252545, 5880, 'Hochbau', None, None, '2022-09-02 00:00:00+02:00', '2023-06-03 00:00:00+02:00', 'Sicherungsmaßnahmen entlang der Straße, Sicherungsmaßnahmen entlang des Gehwegs', 'Erschließung B-Plan Wohngebiet Kiefernweg')
(2, 54.1437060190986, 12.0682192810438, 6050, 'Hochbau', '1c', '1e', '2022-03-07 09:00:00+01:00', '2024-05-31 18:00:00+02:00', 'Sicherungsmaßnahmen entlang der Straße, Sicherungsmaßnahmen entlang des Gehwegs', 'Neubau Mehrfamilienhaus mit Tiefgarage')
(3, 54.0839856179815, 12.1323702387524, 970, 'Gebäudesanierung', '89', '88', '2022-12-07 00:00:00+01:00', '2023-07-01 00:00:00+02:00', None, None)
(4, 54.0775929897237, 12.1167316773101, 2030, 'Hochbau', '24', None, '2022-01-22 00:00:00+01:00', 

Wie wir sehen erhalten wir jede Zeile als Tuple zurück. Dabei kennen wir in diesem Fall nicht die Spaltennamen der Werte. Wir können diese vom Cursor abfragen mit

In [9]:
cur.description

(('index', None, None, None, None, None, None),
 ('latitude', None, None, None, None, None, None),
 ('longitude', None, None, None, None, None, None),
 ('strasse_schluessel', None, None, None, None, None, None),
 ('sparte', None, None, None, None, None, None),
 ('von', None, None, None, None, None, None),
 ('nach', None, None, None, None, None, None),
 ('baubeginn', None, None, None, None, None, None),
 ('bauende', None, None, None, None, None, None),
 ('verkehrsbeeintraechtigungen', None, None, None, None, None, None),
 ('baumassnahme', None, None, None, None, None, None))

Meist fragt man allerdings nicht alle Spalten einer Tabelle mit dem Platzhalter `*` ab, sonder gibt geziehlt die Spalten an, die man erhalten möchte. So kann man unnötigen Datenverkehr spaaren.

In [10]:
sql='SELECT strasse_schluessel, sparte, latitude, longitude, baumassnahme FROM baustellen;'
for row in cur.execute(sql):
    print(row)

(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau')
(5880, 'Hochbau', 54.0675802011562, 12.0763242252545, 'Erschließung B-Plan Wohngebiet Kiefernweg')
(6050, 'Hochbau', 54.1437060190986, 12.0682192810438, 'Neubau Mehrfamilienhaus mit Tiefgarage')
(970, 'Gebäudesanierung', 54.0839856179815, 12.1323702387524, None)
(2030, 'Hochbau', 54.0775929897237, 12.1167316773101, None)
(13290, 'Straßenbau', 54.1143851300814, 12.1025305154084, 'Herstellung Fußwegeverbindung')
(7540, 'Hochbau', 54.0835782924837, 12.1332427590549, 'Neubau MFH')
(7650, 'Straßenbau', 54.0760346216846, 12.1301725386035, 'Herstellung und Neuordnung barrierefreier ZOB Rostock -Baustellenausfahrt-')
(8140, 'Hochbau', 54.0869285867705, 12.1371155734346, None)
(9370, 'Straßenbau', 54.0865062203894, 12.1014245452926, 'Tiefgründiger Ausbau')
(550, 'Hochbau', 54.0829446735607, 12.0865258322934, 'Neubau Institut für Sportwissenschaften')
(6880, 'Hochbau', 54.0941536291858, 12.1011193164597, 'Baustellenausfahrt -B-Plan 

## Filtern von Ergebnissen mit `WHERE`

Oft will man nur bestimmte Zeilen aus einer Tabelle in einer Datenbank abrufen. Diese Auswahl definiert man mit dem SQL Befehl `WHERE` und zusätzlichen logischen Vergleichsoperatoren.

Wollen wir zum Beispiel nur die Baustellen haben die einen Baubegin in 2023 haben, so können wir definieren

In [11]:
sql='''SELECT strasse_schluessel, sparte, latitude, longitude, baumassnahme, baubeginn, bauende
       FROM baustellen
       WHERE baubeginn >= '2023-01-01 00:00:00+01';'''
for row in cur.execute(sql):
    print(row)

(7490, 'Wasserleitung', 54.0844929287006, 12.1133411923645, 'Reparatur Trinkwasserschieber', '2023-01-10 00:00:00+01:00', '2023-02-11 00:00:00+01:00')
(88140, 'Wasserleitung', 54.1442844454467, 12.1047942095532, 'Trinkwasserrohrbruch -Havariemaßnahme-', '2023-01-03 00:00:00+01:00', '2023-01-14 00:00:00+01:00')
(2590, 'Hochbau', 54.0830056344399, 12.128599477197, 'Materiallieferung mit temporärer Vollsperrung', '2023-01-09 07:00:00+01:00', '2023-06-02 16:00:00+02:00')
(1860, 'Wasserleitung', 54.0900581053357, 12.1183804341101, 'Auswechslung Straßenkappe', '2023-01-09 00:00:00+01:00', '2023-01-21 00:00:00+01:00')
(7890, 'Wasserleitung', 54.074414621566, 12.0875195393619, 'Erneuerung Trinkwassergrundstücksanschluss', '2023-01-03 00:00:00+01:00', '2023-01-28 00:00:00+01:00')
(2170, 'Hochbau', 54.0716025252025, 12.1202646686759, None, '2023-01-09 00:00:00+01:00', '2024-03-09 00:00:00+01:00')
(590, 'Kranarbeiten', 54.0910461832592, 12.1462184366184, 'Kranstellung für Balkonmontage an 2 Tagen

Mehrere Bedingungen können durch logische Operatoren wie 'AND', 'OR' oder 'NOT' verknüpft werden. Wollen wir alle Baustellen die im 1 Quartal 2023 angefangen und beendet werden, so schreiben wir

In [12]:
sql='''SELECT strasse_schluessel, sparte, latitude, longitude, baumassnahme, baubeginn, bauende
       FROM baustellen
       WHERE baubeginn >= '2023-01-01 00:00:00+01' AND bauende < '2024-04-01 00:00:00+01';'''
for row in cur.execute(sql):
    print(row)

(7490, 'Wasserleitung', 54.0844929287006, 12.1133411923645, 'Reparatur Trinkwasserschieber', '2023-01-10 00:00:00+01:00', '2023-02-11 00:00:00+01:00')
(88140, 'Wasserleitung', 54.1442844454467, 12.1047942095532, 'Trinkwasserrohrbruch -Havariemaßnahme-', '2023-01-03 00:00:00+01:00', '2023-01-14 00:00:00+01:00')
(2590, 'Hochbau', 54.0830056344399, 12.128599477197, 'Materiallieferung mit temporärer Vollsperrung', '2023-01-09 07:00:00+01:00', '2023-06-02 16:00:00+02:00')
(1860, 'Wasserleitung', 54.0900581053357, 12.1183804341101, 'Auswechslung Straßenkappe', '2023-01-09 00:00:00+01:00', '2023-01-21 00:00:00+01:00')
(7890, 'Wasserleitung', 54.074414621566, 12.0875195393619, 'Erneuerung Trinkwassergrundstücksanschluss', '2023-01-03 00:00:00+01:00', '2023-01-28 00:00:00+01:00')
(2170, 'Hochbau', 54.0716025252025, 12.1202646686759, None, '2023-01-09 00:00:00+01:00', '2024-03-09 00:00:00+01:00')
(590, 'Kranarbeiten', 54.0910461832592, 12.1462184366184, 'Kranstellung für Balkonmontage an 2 Tagen

## Daten Aggregieren

Da die Ergebnisse Zeile für Zeile übertragen werden, ist die Anzahl der Ergebnisse auch nicht unbedingt vorher bekannt. Deshalb unterstützen alle relationale Datenbanken die Aggregatfunktion `count(*)`, um die Anzahl der Zeilen der Ergebnisse zurück zu geben.

In [13]:
sql='SELECT count(*) FROM baustellen;'

for row in cur.execute(sql):
    print(row)

(100,)


Hier erhalten wir nur eine Ergebniszeile zurück, mit der Anzahl der Zeilen in der Tabelle.

Darüber hinaus unerstützt SQL viele Aggregierfunktionen. Wollen wir zum Beispiel in unserem Beispieldatensatz die minimalen Baubeginn aller Baustellen identifizieren nutzen wir die `min()`-Funktion mit dem entsprechenden Spaltennamen.

In [14]:
sql='SELECT min(baubeginn) FROM baustellen;'

for row in cur.execute(sql):
    print(row)

('2020-10-23 00:00:00+02:00',)


Wieder erhalten wir ein Ergebnis mit dem Datum. Das entscheidende ist, dass diese Aggregatfunktionen direkt in der Datenbank ausgeführt wird. Wir müssen also nicht alle Daten runterladen, um einfache Kennwerte wie die Anzahl, das Minimum oder Maximum von Spalten zu erhalten. Dabei können wir auch mehrere Aggregierungen in einer Anfrage und mathematische Berechnungen ausführen. Das folgende Beispiel gibt uns die Anzahl der Baustellen sowie den minimalen Baubeginn und das maximale Bauende sowie die Baudauer. Die Baudauer berechnen wir direkt in der Datenbank als den arithmethischen Durchschnitt `avg` der Differenz zwischen Bauende und Baubeginn in Tagen im Julianischen Kalender.

In [15]:
sql='''SELECT sparte, count(*), min(baubeginn), max(bauende), avg(JulianDay(bauende)-JulianDay(baubeginn))
       FROM baustellen;'''

for row in cur.execute(sql):
    print(row)

('Gleisbau', 100, '2020-10-23 00:00:00+02:00', '2025-01-01 00:00:00+01:00', 339.7066666666651)


## Statistiken Extrahieren mit Aggregieren von Gruppen, Sortieren und Limitieren

Das letzte Beispiel zeigt bereits, wie wir SQL nutzen können um Statistiken der Daten in einer Datenbank zu berechnen. Das wird besonders dann interessant, wenn wir das mit Gruppierungen kombinieren. Gruppierungen fassen die abgefragten Zeilen in der Tabelle entlang den angegebenen Spaltennamen als Gruppe zusammen. Wollen wir zum Beispiel wissen, wie lange im Durchschnitt die Dauer der Baustellen je nach `sparte` ist, so können wir diese Spalte als Gruppe definieren und die mittleren Baudauer für jede Gruppe bestimmen. Dafür bietet SQL den Befehl `GROUP BY` an, hinter dem wir die Spalte angeben nach welcher gruppiert werden soll.

In [16]:
sql='''SELECT sparte, count(*), min(baubeginn), max(bauende), avg(JulianDay(bauende)-JulianDay(baubeginn))
       FROM baustellen 
       GROUP BY sparte;'''

for row in cur.execute(sql):
    print(row)

('Abwasserleitung', 1, '2021-08-17 00:00:00+02:00', '2023-03-01 00:00:00+01:00', 561.0416666669771)
('Fernwärmeleitung', 2, '2022-11-09 00:00:00+01:00', '2023-01-28 00:00:00+01:00', 70.5)
('Gebäudesanierung', 22, '2021-02-01 00:00:00+01:00', '2023-12-24 00:00:00+01:00', 323.2727272727061)
('Gleisbau', 4, '2022-04-19 00:00:00+02:00', '2025-01-01 00:00:00+01:00', 607.5208333334886)
('Hochbau', 29, '2020-10-23 00:00:00+02:00', '2024-07-12 16:00:00+02:00', 507.21695402295103)
('Kabelnetz', 4, '2022-06-15 00:00:00+02:00', '2023-04-01 00:00:00+02:00', 129.75)
('Kranarbeiten', 1, '2023-01-12 00:00:00+01:00', '2023-01-14 00:00:00+01:00', 2.0)
('Lichtsignalanlage', 2, '2022-06-09 00:00:00+02:00', '2024-06-01 00:00:00+02:00', 608.5208333334886)
('Straßenbau', 14, '2022-03-28 00:00:00+02:00', '2023-12-31 00:00:00+01:00', 274.589285714319)
('Stromnetz', 2, '2022-11-18 00:00:00+01:00', '2023-02-01 00:00:00+01:00', 52.0)
('Telefonnetz', 2, '2022-12-01 00:00:00+01:00', '2023-01-28 00:00:00+01:00', 51

Wir sehen jetzt, dass 'Hochbau' die größte Gruppen mit 29 Zeilen (Baustellen) bildet, gefolgt von Straßenbau mit 14 Zeilen. Hochbau braucht Hochbau mit 507 Tagen fast doppelt so lange wie Strassenbau mit 274 Tagen. 

(Da wir die Daten tagesaktuell von Opendata-HRO laden, können sich diese Beispielzahlen im Verlauf der Zeit durchaus ändern).

Um die Ergebnisse zu sortieren bietet SQL den `ORDER BY` Befehl. Wollen wir die Ergebnisse nach der Dauer absteigend (`DESC`) sortieren, so schreiben wir:

In [17]:
sql='''SELECT sparte, count(*), min(baubeginn), max(bauende), avg(JulianDay(bauende)-JulianDay(baubeginn)) as Baudauer 
       FROM baustellen 
       GROUP BY sparte
       ORDER BY Baudauer DESC;'''  # ASC - Aufsteigend, DESC - Absteigend

for row in cur.execute(sql):
    print(row)

('Lichtsignalanlage', 2, '2022-06-09 00:00:00+02:00', '2024-06-01 00:00:00+02:00', 608.5208333334886)
('Gleisbau', 4, '2022-04-19 00:00:00+02:00', '2025-01-01 00:00:00+01:00', 607.5208333334886)
('Abwasserleitung', 1, '2021-08-17 00:00:00+02:00', '2023-03-01 00:00:00+01:00', 561.0416666669771)
('Hochbau', 29, '2020-10-23 00:00:00+02:00', '2024-07-12 16:00:00+02:00', 507.21695402295103)
('Gebäudesanierung', 22, '2021-02-01 00:00:00+01:00', '2023-12-24 00:00:00+01:00', 323.2727272727061)
('Straßenbau', 14, '2022-03-28 00:00:00+02:00', '2023-12-31 00:00:00+01:00', 274.589285714319)
('Wasserleitung', 17, '2021-10-28 00:00:00+02:00', '2024-07-31 00:00:00+02:00', 189.87990196076606)
('Kabelnetz', 4, '2022-06-15 00:00:00+02:00', '2023-04-01 00:00:00+02:00', 129.75)
('Fernwärmeleitung', 2, '2022-11-09 00:00:00+01:00', '2023-01-28 00:00:00+01:00', 70.5)
('Stromnetz', 2, '2022-11-18 00:00:00+01:00', '2023-02-01 00:00:00+01:00', 52.0)
('Telefonnetz', 2, '2022-12-01 00:00:00+01:00', '2023-01-28 00

Sind wir jetzt nur an der Top 3 interessiert, so können wir mit `LIMIT` die Anzahl an Ergebnissen die maximal zurück gegeben werden.

In [18]:
sql='''SELECT sparte, count(*), min(baubeginn), max(bauende), avg(JulianDay(bauende)-JulianDay(baubeginn)) as Baudauer 
       FROM baustellen 
       GROUP BY sparte
       ORDER BY Baudauer DESC
       LIMIT 3;'''

for row in cur.execute(sql):
    print(row)

('Lichtsignalanlage', 2, '2022-06-09 00:00:00+02:00', '2024-06-01 00:00:00+02:00', 608.5208333334886)
('Gleisbau', 4, '2022-04-19 00:00:00+02:00', '2025-01-01 00:00:00+01:00', 607.5208333334886)
('Abwasserleitung', 1, '2021-08-17 00:00:00+02:00', '2023-03-01 00:00:00+01:00', 561.0416666669771)


## Daten aus mehreren Tabellen verknüpfen und gleichzeitig abfragen mit `JOIN`

In unserm Beispiel kennen wir nicht die Straßennamen der Baustelle, sondern nur eine obskure ID namens `strasse_schluessel`. Diese finden wir auch in der Tabelle `adressenliste` zusammen mit dem gesuchten Straßennamen. Wir müssen also nun eine Abfrage über beide Tabellen gemeinsam durchführen, um auch die Straße für die Baustellen zu erfahren. Dafür nutzt man den SQL Befehl `JOIN` und gibt an wo (`WHERE`) die Ergebnisse zusammenzuführen sind indem man eine Gleichheitsbedingung spezifiziert. Da die Spalte `strasse_schluessel` in beiden Tabellen vorkommt und es nicht ganz eindeutig ist auf welche man sich bezieht, gibt man jetzt auch bei den Spaltennamen die Tabelle mit der Dotnotation mit an, also:

In [19]:
sql='''SELECT baustellen.strasse_schluessel, baustellen.sparte, baustellen.latitude, baustellen.longitude, baustellen.baumassnahme, baustellen.baubeginn, baustellen.bauende, 
              adressenliste.gemeindeteil_name, adressenliste.strasse_name 
       FROM baustellen 
       JOIN adressenliste 
       WHERE baustellen.strasse_schluessel=adressenliste.strasse_schluessel;'''
for row in cur.execute(sql):
    print(row)

(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-0

Da diese ständige Wiederholung der Tabellennamen sehr schreiblastig ist, kann man in SQL mit dem Befehl `AS` auch kürzere Namen für Spalten und Tabellen verwenden

In [20]:
sql='''SELECT b.strasse_schluessel, b.sparte, b.latitude, b.longitude, b.baumassnahme, b.baubeginn, b.bauende, 
              a.gemeindeteil_name, a.strasse_name 
       FROM baustellen AS b 
       JOIN adressenliste AS a
       WHERE b.strasse_schluessel=a.strasse_schluessel;'''
for row in cur.execute(sql):
    print(row)

(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-01-01 00:00:00+01:00', 'Stadtmitte', 'Am Strande')
(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', '2022-09-01 00:00:00+02:00', '2024-0

Jetzt kommt es allerdings zu vielen Dopplungen in unseren Ergebnissen, da jede Ergebniszeile aus der Tabelle `baustellen` mit jeder Zeile aus der Tabelle `adressenliste` wiederholt wird, weil in der letzteren ja eine Straße mit dem gleichen `strasse_schluessel` mehrmals vorkommt. Diese Duplikate wollen wir rausfiltern mit dem SQL Befehl `DISTINCT`

In [21]:
sql='''SELECT DISTINCT b.strasse_schluessel, b.sparte, b.latitude, b.longitude, b.baumassnahme, a.gemeindeteil_name, a.strasse_name 
       FROM baustellen AS b
       JOIN adressenliste AS a
       WHERE b.strasse_schluessel=a.strasse_schluessel;'''
for row in cur.execute(sql):
    print(row)

(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', 'Stadtmitte', 'Am Strande')
(5880, 'Hochbau', 54.0675802011562, 12.0763242252545, 'Erschließung B-Plan Wohngebiet Kiefernweg', 'Gartenstadt/Stadtweide', 'Kiefernweg')
(6050, 'Hochbau', 54.1437060190986, 12.0682192810438, 'Neubau Mehrfamilienhaus mit Tiefgarage', 'Groß Klein', 'Kleiner Warnowdamm')
(6050, 'Hochbau', 54.1437060190986, 12.0682192810438, 'Neubau Mehrfamilienhaus mit Tiefgarage', 'Schmarl', 'Kleiner Warnowdamm')
(970, 'Gebäudesanierung', 54.0839856179815, 12.1323702387524, None, 'Stadtmitte', 'Augustenstr.')
(2030, 'Hochbau', 54.0775929897237, 12.1167316773101, None, 'Südstadt', 'Erich-Schlesinger-Str.')
(13290, 'Straßenbau', 54.1143851300814, 12.1025305154084, 'Herstellung Fußwegeverbindung', 'Gehlsdorf', 'Zur Yachtwerft')
(7540, 'Hochbau', 54.0835782924837, 12.1332427590549, 'Neubau MFH', 'Stadtmitte', 'Paulstr.')
(7650, 'Straßenbau', 54.0760346216846, 12.1301725386035, 'Herstellung und Neuordnung barrierefre

Jetzt kennen wir schon den richtigen Straßenname zu jeder Baustelle, allerdings fehlt uns die Gemeinde. Hierfür verknüpfen wir die Daten mit der Tabelle
`gemeinden`. Dafür ergänzen wir die zusätzliche Tabelle im `JOIN` und definieren die weitere Verknüpfungsbedingung auf die Spalte `gemeinde_schluessel`.


In [22]:
sql='''SELECT DISTINCT b.strasse_schluessel, b.sparte, b.latitude, b.longitude, b.baumassnahme, a.gemeindeteil_name, a.strasse_name, g.gemeinde_name
       FROM baustellen AS b
       JOIN adressenliste AS a, gemeinden AS g
       WHERE b.strasse_schluessel=a.strasse_schluessel AND a.gemeinde_schluessel=g.gemeinde_schluessel;'''
for row in cur.execute(sql):
    print(row)

(530, 'Hochbau', 54.0926332355401, 12.1473363033309, 'Hochbau', 'Stadtmitte', 'Am Strande', 'Rostock, Hanse- und Universitätsstadt')
(5880, 'Hochbau', 54.0675802011562, 12.0763242252545, 'Erschließung B-Plan Wohngebiet Kiefernweg', 'Gartenstadt/Stadtweide', 'Kiefernweg', 'Rostock, Hanse- und Universitätsstadt')
(6050, 'Hochbau', 54.1437060190986, 12.0682192810438, 'Neubau Mehrfamilienhaus mit Tiefgarage', 'Groß Klein', 'Kleiner Warnowdamm', 'Rostock, Hanse- und Universitätsstadt')
(6050, 'Hochbau', 54.1437060190986, 12.0682192810438, 'Neubau Mehrfamilienhaus mit Tiefgarage', 'Schmarl', 'Kleiner Warnowdamm', 'Rostock, Hanse- und Universitätsstadt')
(970, 'Gebäudesanierung', 54.0839856179815, 12.1323702387524, None, 'Stadtmitte', 'Augustenstr.', 'Rostock, Hanse- und Universitätsstadt')
(2030, 'Hochbau', 54.0775929897237, 12.1167316773101, None, 'Südstadt', 'Erich-Schlesinger-Str.', 'Rostock, Hanse- und Universitätsstadt')
(13290, 'Straßenbau', 54.1143851300814, 12.1025305154084, 'Herstel

Es gibt [verschiedene Typen von Joins](https://www.w3schools.com/sql/sql_join.asp) je nachdem ob man nur die Ergebnisse haben will für die es:
- Einträge in beiden Tabellen gibt (INNER, Default)
- Einträge in mindestens der ersten Tabelle gibt (RIGHT)
- Einträge in mindestens der zweiten Tabelle gibt (LEFT)
- Einträge in mindestens der zweiten Tabelle gibt (OUTER)

Der `NATURAL JOIN` Typen vereinfachen die Schreibarbeit. Beim `NATURAL JOIN` werden dabei die Spalten gematcht, welche identisch heißen. Er funktioniert somit nur wenn die Spalten in den zu matchenden Tabellen identisch heißen und alle anderen Spalten andere Namen haben, was eher seltener der Fall ist.

In [23]:
sqlNJ='''SELECT DISTINCT strasse_schluessel, sparte, latitude, longitude, baumassnahme, gemeindeteil_name, strasse_name, gemeinde_name
       FROM baustellen
       NATURAL JOIN adressenliste, gemeinden;'''
for row in cur.execute(sqlNJ):
    print(row)

(7400, 'Gleisbau', 54.0804680870093, 12.1265235284217, 'Baustelleneinrichtung / Belieferung für Neubau Goetheplatzbrücke', 'Südstadt', 'Südring', 'Dabel')
(7400, 'Gleisbau', 54.0804680870093, 12.1265235284217, 'Baustelleneinrichtung / Belieferung für Neubau Goetheplatzbrücke', 'Südstadt', 'Südring', 'Sternberg, Stadt')
(7400, 'Gleisbau', 54.0804680870093, 12.1265235284217, 'Baustelleneinrichtung / Belieferung für Neubau Goetheplatzbrücke', 'Südstadt', 'Südring', 'Eldena')
(7400, 'Gleisbau', 54.0804680870093, 12.1265235284217, 'Baustelleneinrichtung / Belieferung für Neubau Goetheplatzbrücke', 'Südstadt', 'Südring', 'Neu Poserin')
(7400, 'Gleisbau', 54.0804680870093, 12.1265235284217, 'Baustelleneinrichtung / Belieferung für Neubau Goetheplatzbrücke', 'Südstadt', 'Südring', 'Strasburg (Uckermark), Stadt')
(7400, 'Gleisbau', 54.0804680870093, 12.1265235284217, 'Baustelleneinrichtung / Belieferung für Neubau Goetheplatzbrücke', 'Südstadt', 'Südring', 'Basedow')
(7400, 'Gleisbau', 54.08046

:::{warning}
Es funktioniert auch in diesem Beispiel nicht. Es werden zwar Ergebnisse zurück gegeben, allerdings, deutlich mehr als im letzten Join. Die Gemeinden werden hier kreuzkombiniert. Man sieht also, dass der `NATURAL JOIN` mit Vorsicht zu benutzen ist.
:::

## Daten mit SQL und Pandas abfragen und mit GeoJSON visualisieren

Wir haben am Anfang ja die Tabellen in der SQLite Datenbank bereits mit Pandas erzeugt. Wir können Pandas auch nutzen um Daten aus einer SQL Anfrage direkt als Dataframe (Pandas Tabelle) zu laden. 

Fügen wir einmal die Beispiele zusammen und lassen uns alle Baustellen mit Strasse und Gemeinde im ersten Quartal 2023 auf einer Karte anzeigen. Die SQL Abfrage dazu sieht wie folgt aus

In [24]:
sql='''SELECT DISTINCT b.strasse_schluessel, b.sparte, b.latitude, b.longitude, b.baumassnahme, b.baubeginn, b.bauende, a.gemeindeteil_name, a.strasse_name, g.gemeinde_name
       FROM baustellen AS b
       JOIN adressenliste AS a, gemeinden AS g
       WHERE b.strasse_schluessel=a.strasse_schluessel AND a.gemeinde_schluessel=g.gemeinde_schluessel
         AND b.baubeginn >= '2023-01-01 00:00:00+01' AND b.bauende < '2024-04-01 00:00:00+01';'''

Wir führen die SQL Abfrage direkt mit der Pandas Funktion `pd.read_sql(sql, con)` aus und erzeugen uns einen DataFrame der die gewünschten Daten enthält.

In [25]:
baustellen_mit_strasse = pd.read_sql(sql, con)
baustellen_mit_strasse

strasse_schluessel            sparte   latitude  longitude  \
0                 7490     Wasserleitung  54.084493  12.113341   
1                 7490     Wasserleitung  54.084493  12.113341   
2                88140     Wasserleitung  54.144284  12.104794   
3                 2590           Hochbau  54.083006  12.128599   
4                 2590           Hochbau  54.083006  12.128599   
5                 1860     Wasserleitung  54.090058  12.118380   
6                 7890     Wasserleitung  54.074415  12.087520   
7                 2170           Hochbau  54.071603  12.120265   
8                  590      Kranarbeiten  54.091046  12.146218   
9                 1390     Wasserleitung  54.141787  12.045579   
10                6550     Wasserleitung  54.088715  12.119047   
11               11210         Stromnetz  54.149454  12.174907   
12                7090     Wasserleitung  54.085126  12.150060   
13                7090     Wasserleitung  54.085126  12.150060   
14                6720           Hochbau  54.093835  12.092313   
15                1140  Gebäudesanierung  54.095053  12.092238   
16                3440        Straßenbau  54.092537  12.087238   
17                8180           Hochbau  54.076420  12.096935   
18                8180           Hochbau  54.076420  12.096935   
19                8180           Hochbau  54.076420  12.096935   
20                7520  Gebäudesanierung  54.091470  12.118661   

                                     baumassnahme                  baubeginn  \
0                   Reparatur Trinkwasserschieber  2023-01-10 00:00:00+01:00   
1                   Reparatur Trinkwasserschieber  2023-01-10 00:00:00+01:00   
2          Trinkwasserrohrbruch -Havariemaßnahme-  2023-01-03 00:00:00+01:00   
3   Materiallieferung mit temporärer Vollsperrung  2023-01-09 07:00:00+01:00   
4   Materiallieferung mit temporärer Vollsperrung  2023-01-09 07:00:00+01:00   
5                       Auswechslung Straßenkappe  2023-01-09 00:00:00+01:00   
6      Erneuerung Trinkwassergrundstücksanschluss  2023-01-03 00:00:00+01:00   
7                                            None  2023-01-09 00:00:00+01:00   
8       Kranstellung für Balkonmontage an 2 Tagen  2023-01-12 00:00:00+01:00   
9        Reparatur Einfallstelle Gehweg (Havarie)  2023-01-03 00:00:00+01:00   
10                     Einbau Trinkwasserschieber  2023-01-11 00:00:00+01:00   
11                Hausanschluss Strom -1 Werktag-  2023-01-03 00:00:00+01:00   
12  Auswechselung Wasserzählerschacht -1 Werktag-  2023-01-03 00:00:00+01:00   
13  Auswechselung Wasserzählerschacht -1 Werktag-  2023-01-03 00:00:00+01:00   
14                                           None  2023-01-09 00:00:00+01:00   
15                                           None  2023-01-10 00:00:00+01:00   
16                                Gehwegsanierung  2023-01-09 00:00:00+01:00   
17                                           None  2023-01-03 00:00:00+01:00   
18                                           None  2023-01-03 00:00:00+01:00   
19                                           None  2023-01-03 00:00:00+01:00   
20                                           None  2023-01-09 00:00:00+01:00   

                      bauende        gemeindeteil_name          strasse_name  \
0   2023-02-11 00:00:00+01:00             Hansaviertel              Parkstr.   
1   2023-02-11 00:00:00+01:00  Kröpeliner-Tor-Vorstadt              Parkstr.   
2   2023-01-14 00:00:00+01:00                     Peez         Ost-West-Str.   
3   2023-06-02 16:00:00+02:00  Kröpeliner-Tor-Vorstadt        St.-Georg-Str.   
4   2023-06-02 16:00:00+02:00               Stadtmitte        St.-Georg-Str.   
5   2023-01-21 00:00:00+01:00  Kröpeliner-Tor-Vorstadt        Doberaner Str.   
6   2023-01-28 00:00:00+01:00   Gartenstadt/Stadtweide         Rennbahnallee   
7   2024-03-09 00:00:00+01:00                 Südstadt    Erich-Weinert-Str.   
8   2023-01-14 00:00:00+01:00               Stadtmitte                


Die Tabelle enthält ja bereits die Latitude und Longitude. Mit dem Paket `pandas_geojson` können wir aus den Dataframe in ein GeoJSON FeatureCollection umwandeln. Wir installieren zuerst das Paket mit pip.

In [26]:
pip install pandas_geojson --quiet

Note: you may need to restart the kernel to use updated packages.


In [27]:
from pandas_geojson import to_geojson

geo_json = to_geojson(df=baustellen_mit_strasse, lat='latitude', lon='longitude',
                 properties=['strasse_name','sparte','baumassnahme', 'baubeginn', 'bauende'])

Jetzt können wir uns die Baustellen  mit dem schon bekannten Paket `geojsonio` auf einer Karte visualisieren.

In [28]:
import json
import geojsonio

geojsonio.display(json.dumps(geo_json))

'http://geojson.io/#data=data:application/json,%7B%22type%22%3A%20%22FeatureCollection%22%2C%20%22features%22%3A%20%5B%7B%22type%22%3A%20%22Feature%22%2C%20%22properties%22%3A%20%7B%22strasse_name%22%3A%20%22Parkstr.%22%2C%20%22sparte%22%3A%20%22Wasserleitung%22%2C%20%22baumassnahme%22%3A%20%22Reparatur%20Trinkwasserschieber%22%2C%20%22baubeginn%22%3A%20%222023-01-10%2000%3A00%3A00%2B01%3A00%22%2C%20%22bauende%22%3A%20%222023-02-11%2000%3A00%3A00%2B01%3A00%22%7D%2C%20%22geometry%22%3A%20%7B%22type%22%3A%20%22Point%22%2C%20%22coordinates%22%3A%20%5B12.1133411923645%2C%2054.0844929287006%5D%7D%7D%2C%20%7B%22type%22%3A%20%22Feature%22%2C%20%22properties%22%3A%20%7B%22strasse_name%22%3A%20%22Parkstr.%22%2C%20%22sparte%22%3A%20%22Wasserleitung%22%2C%20%22baumassnahme%22%3A%20%22Reparatur%20Trinkwasserschieber%22%2C%20%22baubeginn%22%3A%20%222023-01-10%2000%3A00%3A00%2B01%3A00%22%2C%20%22bauende%22%3A%20%222023-02-11%2000%3A00%3A00%2B01%3A00%22%7D%2C%20%22geometry%22%3A%20%7B%22type%22%3A%20

![Baustellen](images/baustellen.png)